File: src/chapters/differential-equations/content/code/heun.py
def heun(f, t0, y0, h, n):
    """Heun's method (RK2, explicit trapezoidal rule)."""
    t, y = t0, y0
    for _ in range(n):
        k1 = f(t, y)
        k2 = f(t + h, y + h * k1)
        y += h * (k1 + k2) / 2
        t += h
    return t, y


if __name__ == "__main__":
    f = lambda t, y: 2 * y - 10 * t ** 2 + 2 * t
    print("y(1) =", heun(f, 0.0, 1.0, 0.1, 10)[1])

File: src/chapters/differential-equations/content/code/heun.cpp
#include <iostream>
#include <functional>
using namespace std;

// Heun's method (RK2, explicit trapezoidal rule).
double heun(function<double(double, double)> f, double t0, double y0, double h, int n) {
    double t = t0, y = y0;
    for (int i = 0; i < n; ++i) {
        double k1 = f(t, y), k2 = f(t + h, y + h * k1);
        y += h * (k1 + k2) / 2; t += h;
    }
    return y;
}

int main() {
    auto f = [](double t, double y) { return 2 * y - 10 * t * t + 2 * t; };
    cout << "y(1) = " << heun(f, 0.0, 1.0, 0.1, 10) << "\n";
}
